docs/features/source-generators.md

```diff
@@ -0,0 +1,291 @@
+`.cshtml` file). 
+
+We discussed embedding this directly in the source text via `#pragma` but 
+this would require language changes and limit the feature to a specific version
+of C#. Other considerations could be specially formed comments or `#if FALSE --`
+blocks. In general a 'side-channel' approach seems preferable to specially crafted
+grammar in the generated text. 
+
+This is not necessarily a goal required for the success of Source Generators;
+Razor’s language service can be updated to work with source generators 
+if it proves to be infeasible, but it certainly something we want to consider 
+as part of the work.
+
+### MSBuild Integration
+
+It is expected that generators will need some form of configuration system, and we intend to allow 
+certain properties to flow through from MSBuild to facilitate this.
+
+> **Note**: This is still under design and open to change.
+
+
+### Performance targets
+
+Ultimately, the performance of the feature is going to be somewhat dependent on the performance of the 
+generators written by customers. Progressive opt-in, and build-time only by-default will allow the IDE
+to mitigate many of the potential performance problems posed by third party generators. However, there
+is still a risk that third-party generators will cause unacceptable performance problems for the IDE, 
+and the design of the feature will need to keep this in mind.
+
+For 1st party generators, especially Razor and Blazor, we aim at a minimum to match the existing
+performance seen by users today. It is expected that even naïve generator-based implementations 
+will perform significantly faster than the existing tooling, due to less communication overhead
+and duplicated work, but improving the speed of these experiences is not a primary goal of this project.
+
+### Language Changes
+
+This design does not currently propose altering the language, it is purely a compiler feature. 
+The previous design for source generators introduced the `replace` and `original` keywords. 
+This proposal removes these, as the source generated is purely additional and so there is no
+need for them. We expect that most scenarios are possible with the existing use of `partial` 
+definitions; as a V1 we expect to ship in this state. If concrete scenarios are later shown
+that can’t be achieved with the V1 approach we would consider allowing modification as a V2. 
+
+## Use cases
+
+We've identified several first and third party candidates that would benefit from source generators:
+
+- ASP.Net: Improve startup time 
+- Blazor and Razor: Massively reduce tooling burden 
+- Azure Functions: regex compilation during startup
+- Azure SDK
+- [gRPC](https://docs.microsoft.com/en-us/aspnet/core/grpc/?view=aspnetcore-3.1)
+- Resx file generation
+- [System.CommandLine](https://github.com/dotnet/command-line-api)
+- Serializers
+- [SWIG](http://www.swig.org/)
+
+
+## Discussion / Open Issues / TODOs:
```

Comment:
The high level design goals mentions that unordered execution is out of scope. A `DependsOn` / `Before` / `After` model is very important as generators very frequently need to depend on the generated code of other generators. If that may help, here's how we implemented it in [`Uno.SourceGeneration`](https://github.com/unoplatform/Uno.SourceGeneration#general-guidelines-for-creating-generators), it's using literal strings to enable that as we haven't found a better way to do that.